src/main/scala/com/github/codelionx/dodo/discovery/CandidateGenerator.scala

```diff
@@ -0,0 +1,17 @@
+package com.github.codelionx.dodo.discovery
+
+import scala.collection.immutable.Queue
+
+
+trait CandidateGenerator {
+  def generateFirstCandidates(columns: Set[Int]): Queue[(List[Int], List[Int])] = {
+    val fixedOrderCols = columns.toList
+    val candidates = fixedOrderCols.flatMap(col1 => fixedOrderCols.drop(fixedOrderCols.indexOf(col1) + 1).map(col2 => (List(col1), List(col2))))
+    Queue(candidates: _*)
+  }
+
+  def generateODCandidates(columns: Set[Int], od: (List[Int], List[Int])): Queue[(List[Int], List[Int])] = {
+    val newCandidates = columns.map(col => (od._1 :+ col, od._2))
```

Comment:
ALGORITHM3 in the paper does only consider the set `A+` as possible attributes to add to the list as the next candidate. `A+` is `columns` minus `od._1.toSet` and minus `od._2.toSet`.